packages/grafana-data/src/transformations/transformers/groupBy.ts

```diff
@@ -0,0 +1,179 @@
+import { DataTransformerID } from './ids';
+import { DataFrame, FieldType, Field } from '../../types/dataFrame';
+import { DataTransformerInfo } from '../../types/transformations';
+import { getFieldDisplayName } from '../../field/fieldState';
+import { ArrayVector } from '../../vector/ArrayVector';
+import { guessFieldTypeForField } from '../../dataframe/processDataFrame';
+import { reduceField, ReducerID } from '../fieldReducer';
+
+export interface GroupByTransformerOptions {
+  byField: string | null;
+  calculationsByField: Array<[string | null, ReducerID[]]>;
```

Comment:
this needs to change so we can support multiple group by fields. 

The tuple array type is a bit hard work with. think something like this could be easier and more readable
```
export interface GroupByTransformerOptions {
 fields: Record<string, GroupByFieldOptions>
}

export interface GroupByFieldOptions {
 aggregations: ReducerId[];
 groupBy: boolean; 
}
```

So just single map for field name -> options, and each option object can have a groupBy Boolean to true/false (if true aggregations is ignored/empty)

Reply:
I agree with @torkelo regarding the structure of the `calculationsByField`. Might be worth having a look at this before we merge this one since it will be stored in the json model of the dashboard.

Reply:
Intuitively I would set `fields` to an Array instead of a Record considering how the UI and the options would be added/removed (on a per line basis) and also how the transform needs to interpret the configuration (builds arrays out of the config options, which is slightly easier starting from an Array, though we agree it can be done fairly easily with both types).

```
export interface GroupByTransformerOptions {
 fields:  GroupByFieldOptions[];
}

export interface GroupByFieldOptions {
 fieldName: string;
 aggregations: ReducerId[];
 groupBy: boolean; 
}
```

Do you see any disadvantage to use the above structure instead ? @mckn @torkelo

Reply:
So one advantage of using a `Record<string, GroupByFieldOptions>` is that you don't need to loop through the `options.fields` list to find the field. You can easily just get it by using `options.fields[fieldName]`.

But at the same time this will be pretty small array so I don't think it will have any noticeable benefits. I think you can use your suggested structure above. It makes it easier to argue and read the code then the existing tuple array.

Reply:
The benefits are more in diffs and merges and how it's stored, we sort all props on alphabetical order

Reply:
@torkelo great input! I wasn't aware of that 🙌

Reply:
So to explain it a bit more in detail. The transformation configuration will be stored in the dashboard `json` data structure that represents the dashboard when it is persisted (saved). When doing changes/comparing this structure it is better to work with a record because of how we store the data.

Reply:
So currently I'm using the following structure:

```
export enum GroupByOperationID {
  aggregate = 'aggregate',
  groupBy = 'groupby',
}

export interface GroupByFieldOptions {
  fieldName: string | null;
  aggregations: ReducerID[];
  operation: GroupByOperationID;
}

export interface GroupByTransformerOptions {
  fields: Record<string, GroupByFieldOptions>;
  fieldsArray: GroupByFieldOptions[];
}
```

It is much easier to work with an array considering how the UI is made, but if we really prefer a Record, I can convert it to a Record to store it in the transform options and convert it back to an Array when needed. This way it would be stored as a Record in the options.

The main thing to consider here is the ordering. Currently I am assuming the user would set a certain order that he could want to keep in the fields he wants to aggregate, which would not really be guaranteed if we store the info in a Record, but the ordering could be dealed with the Organise Transform instead, so we could just sort it however we want and the user can reorganize it afterwards.

What do you think ? How should we sort in the UI ? @torkelo @mckn